common/cuda_hip/components/atomic.hpp

```diff
@@ -95,15 +95,65 @@ __forceinline__ __device__ ResultType reinterpret(ValueType val)
         }                                                                    \
     };
 
+
+#define GKO_BIND_ATOMIC_HELPER_FAKE_STRUCTURE(CONVERTER_TYPE)               \
+    template <typename ValueType>                                           \
+    struct atomic_helper<                                                   \
+        ValueType,                                                          \
+        std::enable_if_t<(sizeof(ValueType) == sizeof(CONVERTER_TYPE))>> {  \
+        __forceinline__ __device__ static ValueType atomic_add(             \
+            ValueType* __restrict__ addr, ValueType val)                    \
+        {                                                                   \
+            assert(false);                                                  \
+            using c_type = CONVERTER_TYPE;                                  \
+            return atomic_wrapper(                                          \
+                addr, [&val](c_type& old, c_type assumed, c_type* c_addr) { \
+                    old = *c_addr;                                          \
+                    *c_addr = reinterpret<c_type>(                          \
```

Comment:
It is generalized from the code in cuda documentation, so I assume it is not undefined behavior from cuda perspective.
Ignoring the UB, I prefer the `reinterpret_cast` rather than `memcpy` because it does not copy the data for sure.
`memcpy` will be up to compiler optimization to copy or reinterpret-like action.

Reply:
Ok, perhaps the device code has stronger UB guarantees. Fine with me.

Reply:
I can tell you as much as this: It is definitely UB, but also the nvcc compiler is unlikely to do optimizations that break it (quote unnamed NVIDIA person). That being said, IMO the worry that the compiler might not be able to optimize such a memcpy is unfounded. That's one of the most basic optimizations it can do here.

Reply:
Not sure about the optimization.
For example, we have used `std::memcpy` in hip device code https://github.com/ginkgo-project/ginkgo/blob/7b0409498bab8e89f9fb232890be8df04548b32f/hip/components/memory.hip.hpp#L102
It is not the constexpr code, so we should not be able to use it in device code.
I assume compiler can optimize it such that avoid the host code issue.
However, it sometimes fails and shows `std::memcpy` is not allowed in device code (replaced by memcpy to fix), does it means that the compiler can not really optimize memcpy in some cases?

Reply:
This has nothing to do with optimization, it's about whether std::memcpy and memcpy are different functions or not. [libstdc++ defines them as a `using` alias](https://github.com/gcc-mirror/gcc/blob/master/libstdc%2B%2B-v3/include/c_std/cstring), which is why std::memcpy often works (can't tell you exactly why it doesn't always work, but that doesn't matter here)
GPU compilers are very aggressive in their memory-to-register promotion, which is what gets rid of this memcpy.